peers.py

```diff
@@ -42,3 +44,44 @@ def get_nodes(self):
             nodes.append((conv.scope, conv.get_remote('private-address')))
 
         return nodes
+
+    def restarted_nodes(self):
+        nodes = []
+        nonce = leader_get('restart_nonce')
+        if not nonce:
+            return nodes  # We're not restarting if no nonce is set.
+        for conv in self.conversations():
+            if conv.get_remote('restarted.{}'.format(nonce)):
+                nodes.append((conv.scope, conv.get_remote('private-address')))
+
+        return nodes
+
+    def set_zk_leader(self):
+        '''
+        Inform peers that the unit that calls this method is the Zookeeper leader.
+
+        Note that Zookeeper tracks leadership separately from juju;
+        the Zookeeper leader is not necessarily the Juju leader.
+
+        '''
+        for conv in self.conversations():
```

Comment:
Docs on Zookeeper leader election are here: http://zookeeper.apache.org/doc/current/recipes.html#sc_leaderElection

Per convo in Hangout w/ @ktsakalozos, we do not currently handle the case where a Zookeeper leader election happens without triggering an event that Juju is wired to notice (e.g. the process crashes, but the machine does not go away). If that happens, Juju may have the wrong node flagged as the Zookeeper leader.

Unfortunately, there really isn't an obvious way to handle that. Juju reactive handlers trigger when a Juju event happens; by definition, they can't trigger when something happens that Juju doesn't register as an event.

We've decided that the best thing to do is to document this possibility, and open a ticket to deal with it if it turns out to be something that happens a significant number of times in the field. (An operator could fix things in the meantime by doing a manual rolling restart on the cluster).